### PR TITLE
implement game pause/play feature

### PR DESCRIPTION
**Pull request by Bashar-Ahmed**

### CODE REVIEW

File: game/game.app.json [hunk 1]
```diff
@@ -8,7 +8,7 @@
         "title": "Rootex Splash",
         "width": 400
     },
-    "startScene": "game/assets/scenes/mesh_collision.scene.json",
+    "startScene": "game/assets/scenes/physics.scene.json",
```
Comment (Ishan-002) [inline game/game.app.json:11]: Change this to mesh_collision

File: game/game.app.json [hunk 2]
```diff
@@ -37,8 +50,8 @@
         },
         "PhysicsSystem": {},
         "UISystem": {
-            "height": 1387,
-            "width": 2560
+            "height": 801,
+            "width": 1536
```
Comment (Ishan-002) [inline game/game.app.json:54]: Revert the height and width

File: rootex/framework/entity.cpp [hunk 3]
```diff
@@ -175,6 +175,7 @@ bool Entity::setScriptJSON(const JSON::json& script)
 		if (OS::IsExists(script["path"]))
 		{
 			m_Script.reset(new Script(script));
+			m_Script->setup(this);
```
Comment (Ishan-002) [inline rootex/framework/entity.cpp:178]: Why is this needed here?

File: rootex/framework/scene.cpp [hunk 4]
```diff
@@ -302,6 +304,13 @@ void Scene::setName(const String& name)
 {
 	m_Name = name;
 	m_FullName = name + " # " + std::to_string(m_ID);
+	m_ScenePause = false;
```
Comment (Ishan-002) [inline rootex/framework/scene.cpp:307]: Why is pausing the scene needed here?

File: rootex/framework/scene.cpp [hunk 5]
```diff
@@ -302,6 +304,13 @@ void Scene::setName(const String& name)
...
+}
+
+void Scene::setScenePause(bool pause)
+{
+	m_ScenePause = pause;
+	return;
```
Comment (Ishan-002) [inline rootex/framework/scene.cpp:313]: remove ```return```

File: rootex/framework/system.cpp [hunk 6]
```diff
@@ -52,6 +52,27 @@ void System::setActive(bool enabled)
...
 
+void System::pause()
+{
+	for (auto systemType : GetSystems())
+	{
+		for (auto system : systemType)
+		{
+			system->m_SystemPause = true;
+		}
+	}
```
Comment (Ishan-002) [inline rootex/framework/system.cpp:63]: Use ```auto&``` instead of ```auto```

File: rootex/framework/system.h [hunk 7]
```diff
@@ -35,6 +35,7 @@ class System
 	String m_SystemName;
 	UpdateOrder m_UpdateOrder;
 	bool m_IsActive;
+	bool m_SystemPause = false;
```
Comment (Ishan-002) [inline rootex/framework/system.h:38]: Initialise the variable in the contructor, not here.

File: rootex/framework/system.h [hunk 8]
```diff
@@ -55,6 +56,8 @@ class System
 	bool isActive() const { return m_IsActive; }
 
 	void setActive(bool enabled);
+	static void pause();
+	static void unpause();
```
Comment (Ishan-002) [inline rootex/framework/system.h:60]: ```unpause``` -> ```unPause```

File: rootex/framework/systems/audio_system.cpp [hunk 9]
```diff
@@ -109,21 +109,51 @@ void AudioSystem::begin()
 
 void AudioSystem::update(float deltaMilliseconds)
 {
-	ZoneScoped;
-
-	for (auto& mc : ECSFactory::GetAllMusicComponent())
+	if (!(m_SystemPause))
```
Comment (Ishan-002) [inline rootex/framework/systems/audio_system.cpp:112]: ``m_SystemPause`` -> ```m_IsSystemPaused```

File: rootex/framework/scene.h [hunk 10]
```diff
@@ -36,6 +36,8 @@ class Scene
 	};
 
 private:
+	bool m_ScenePause = true;
```
Comment (Ishan-002) [inline rootex/framework/scene.h:39]: ```m_ScenePause``` -> ```m_IsScenePaused```

File: rootex/framework/systems/audio_system.cpp [hunk 11]
```diff
@@ -109,21 +109,51 @@ void AudioSystem::begin()
...
-	{
-		m_Listener->update();
+		for (auto& mc : ECSFactory::GetAllMusicComponent())
+		{
+			if (mc.getOwner().getScene()->getScenePause())
+			{
+				continue;
+			}
+			mc.getAudioSource()->queueNewBuffers();
+			mc.update();
```
Comment (Ishan-002) [inline rootex/framework/systems/audio_system.cpp:142]: Write it as 
``` 
if (!mc.getOwner().getScene()->getScenePause())
{
    // update
}
```

File: rootex/framework/systems/audio_system.cpp [hunk 12]
```diff
@@ -109,21 +109,51 @@ void AudioSystem::begin()
...
+		}
+		for (auto& smc : ECSFactory::GetAllShortMusicComponent())
+		{
+			if (smc.getOwner().getScene()->getScenePause())
+			{
+				continue;
+			}
+			smc.update();
```
Comment (Ishan-002) [inline rootex/framework/systems/audio_system.cpp:150]: same

File: rootex/framework/systems/script_system.cpp [hunk 13]
```diff
@@ -62,30 +62,58 @@ void CallUpdateForScene(Scene* scene, float deltaMilliseconds)
...
+			{
+				entity->call("enterScene", { entity });
+			}
 		}
+		m_ScriptEntitiesToEnter.clear();
+
+		Scene* root = SceneLoader::GetSingleton()->getRootScene();
+		CallUpdateForScene(root, deltaMilliseconds);
+
+		m_FluxTweener["update"](deltaMilliseconds * MS_TO_S);
```
Comment (Ishan-002) [inline rootex/framework/systems/script_system.cpp:89]: This part is common in both if and else. You can put it outside the conditional statements.

File: editor/gui/inspector_dock.cpp [hunk 14]
```diff
@@ -211,6 +211,7 @@ void InspectorDock::draw(float deltaMilliseconds)
 						}
 					}
 				}
+				ImGui::Checkbox("Stop work on Pause", &m_OpenedScene->getScenePause());
```
Comment (Ishan-002) [inline editor/gui/inspector_dock.cpp:214]: What does "Stop work on pause" mean?

File: rootex/framework/entity.cpp (hunk 3 above)
Comment (Bashar-Ahmed) [reply rootex/framework/entity.cpp:178]: Initially Exception was being thrown on running a scene after adding child scene from file, because for the scenes added from file, `setup` was not being called.

File: rootex/framework/scene.cpp (hunk 4 above)
Comment (Bashar-Ahmed) [reply rootex/framework/scene.cpp:307]: Not needed ig.
Will correct it.

File: editor/gui/inspector_dock.cpp (hunk 14 above)
Comment (Bashar-Ahmed) [reply editor/gui/inspector_dock.cpp:214]: Stop the systems of that scene during `pause` .
Renaming it to

File: rootex/framework/systems/pause_system.cpp [hunk 15]
```diff
@@ -0,0 +1,38 @@
+#include "pause_system.h"
+
+#define LINE_MAX_VERTEX_COUNT 1000
+#define LINE_VERTEX_COUNT 2
```
Comment (sin3point14) [inline rootex/framework/systems/pause_system.cpp:4]: used nowhere

File: rootex/framework/systems/pause_system.cpp [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+		if (!(Scene::FindScenesByName("PauseScene").size()))
+		{
+			Ptr<Scene>& pauseScene = Scene::CreateFromFile("game/assets/scenes/pause.scene.json");
+			pauseScene->setName("PauseScene");
+			root->addChild(pauseScene);
+		}
+	}
+	else
+	{
+		Vector<Scene*> pauseScenes = Scene::FindScenesByName("PauseScene");
```
Comment (sin3point14) [inline rootex/framework/systems/pause_system.cpp:32]: Open an issue to create a mechanism for Reserved Entity Names, so that a player won't be able to create entities that interfere with internal engine mechanisms

File: rootex/script/interpreter.cpp [hunk 17]
```diff
@@ -310,6 +325,33 @@ void LuaInterpreter::registerTypes()
...
+			sol::usertype<PlayerSystem> playerSystem = systems.new_usertype<PlayerSystem>("Player", sol::base_classes, sol::bases<System>());
+			sol::usertype<PostProcessSystem> postProcessSystem = systems.new_usertype<PostProcessSystem>("PostProcess", sol::base_classes, sol::bases<System>());
+			sol::usertype<RenderSystem> renderSystem = systems.new_usertype<RenderSystem>("Render", sol::base_classes, sol::bases<System>());
+			sol::usertype<RenderUISystem> renderUISystem = systems.new_usertype<RenderUISystem>("RenderUI", sol::base_classes, sol::bases<System>());
+			sol::usertype<ScriptSystem> scriptSystem = systems.new_usertype<ScriptSystem>("Script", sol::base_classes, sol::bases<System>());
+			sol::usertype<TransformAnimationSystem> transformAnimationSystem = systems.new_usertype<TransformAnimationSystem>("TransformAnimation", sol::base_classes, sol::bases<System>());
+			sol::usertype<TriggerSystem> triggerSystem = systems.new_usertype<TriggerSystem>("Trigger", sol::base_classes, sol::bases<System>());
+			sol::usertype<UISystem> uiSystem = systems.new_usertype<UISystem>("UI", sol::base_classes, sol::bases<System>());
```
Comment (sin3point14) [inline rootex/script/interpreter.cpp:349]: these are type definitions, rename them to by of the kind: `sol::usertype<AnimationSystem> animationSystem = systems.new_usertype<AnimationSystem>("AnimationSystem", sol::base_classes, sol::bases<System>());`
and then add property access like this:
```
systems["Animation"] = sol::property(&AnimationSystem::GetSingleton)
```

File: rootex/script/interpreter.cpp [hunk 18]
```diff
@@ -310,6 +325,33 @@ void LuaInterpreter::registerTypes()
...
+
+			pauseSystem["getIsPausingEnabled"] = []() { return PauseSystem::GetSingleton()->getIsPausingEnabled(); };
+			pauseSystem["setIsPausingEnabled"] = [](bool pausing) { PauseSystem::GetSingleton()->setIsPausingEnabled(pausing); };
```
Comment (sin3point14) [inline rootex/script/interpreter.cpp:352]: after adding property access to call GetSingleton, this would change to
```
pauseSystem["getIsPausingEnabled"] = &PauseSystem::getIsPausingEnabled
```

**Review by sin3point14 — CHANGES_REQUESTED** [review]
Also remove the excessive code duplication we have in systems

**Review by sin3point14 — APPROVED** [review]
dorime 🙏
